fast path for declared_generic_bounds_from_env

aliemjay opened the pull request:
~2% perf gain for diesel

aliemjay committed:
fast path for declared_generic_bounds_from_env

aliemjay commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 20e222a73065f8ec6657aedd3045f44ebacc176d with merge e0a7b59fd81315e056574a0f8fb51546821d3753...
<!-- homu: {"type":"TryBuildStarted","head_sha":"20e222a73065f8ec6657aedd3045f44ebacc176d","merge_sha":"e0a7b59fd81315e056574a0f8fb51546821d3753"} -->

compiler-errors reviewed (approved):
r=me after perf

compiler-errors commented on compiler/rustc_infer/src/infer/outlives/verify.rs:
```diff
@@ -245,10 +245,19 @@ impl<'cx, 'tcx> VerifyBoundCx<'cx, 'tcx> {
                 );
+                // Fast path for the common case.
+                match (erased_ty.kind(), &p) {
+                    (ty::Alias(_, a1), GenericKind::Alias(a2)) if a1.def_id != a2.def_id => {
```
This *could* use a comment saying that we expect all preds here to have only rigid projections.

aliemjay commented on compiler/rustc_infer/src/infer/outlives/verify.rs:
```diff
@@ -245,10 +245,19 @@ impl<'cx, 'tcx> VerifyBoundCx<'cx, 'tcx> {
                 );
+                // Fast path for the common case.
+                match (erased_ty.kind(), &p) {
+                    (ty::Alias(_, a1), GenericKind::Alias(a2)) if a1.def_id != a2.def_id => {
+                        return None;
+                    }
+                    (ty::Param(p1), GenericKind::Param(p2)) if p1 != p2 => return None,
+                    _ => {}
```
```suggestion
                    (ty::Alias(_, a1), GenericKind::Alias(a2)) if a1.def_id == a2.def_id =>{}
                    (ty::Param(p1), GenericKind::Param(p2)) if p1 == p2 => {}
                    _ => return None,
```
This should cover more cases.

aliemjay replied:
This is missing the case of `GenericKind::Placeholder`.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7250973431/job/19755350614)
Build commit: e0a7b59fd81315e056574a0f8fb51546821d3753 (`e0a7b59fd81315e056574a0f8fb51546821d3753`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7250973431/job/19755350614"},"merge_sha":"e0a7b59fd81315e056574a0f8fb51546821d3753"} -->

rust-timer commented:
Queued e0a7b59fd81315e056574a0f8fb51546821d3753 with parent a7690a366ccb7c7fd6865241792e2199b23c2b5a, future [comparison URL](https://perf.rust-lang.org/compare.html?start=a7690a366ccb7c7fd6865241792e2199b23c2b5a&end=e0a7b59fd81315e056574a0f8fb51546821d3753).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.0 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (e0a7b59fd81315e056574a0f8fb51546821d3753): [comparison URL](https://perf.rust-lang.org/compare.html?start=a7690a366ccb7c7fd6865241792e2199b23c2b5a&end=e0a7b59fd81315e056574a0f8fb51546821d3753&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=a7690a366ccb7c7fd6865241792e2199b23c2b5a&end=e0a7b59fd81315e056574a0f8fb51546821d3753&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.4% | [-2.2%, -0.2%] | 8     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.4% | [-2.2%, -0.2%] | 8     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a7690a366ccb7c7fd6865241792e2199b23c2b5a&end=e0a7b59fd81315e056574a0f8fb51546821d3753&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.5%, 0.5%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 1.3%  | [1.3%, 1.4%]   | 2     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.5%, -0.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.0% | [-0.5%, 0.5%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=a7690a366ccb7c7fd6865241792e2199b23c2b5a&end=e0a7b59fd81315e056574a0f8fb51546821d3753&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.4%, 0.7%]   | 4     |
| Regressions ❌ <br /> (secondary)  | 2.1%  | [1.6%, 2.5%]   | 5     |
| Improvements ✅ <br /> (primary)   | -1.3% | [-1.5%, -1.1%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.1% | [-1.5%, 0.7%]  | 6     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=a7690a366ccb7c7fd6865241792e2199b23c2b5a&end=e0a7b59fd81315e056574a0f8fb51546821d3753&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 673.256s -> 671.287s (-0.29%)
**Artifact size**: 312.47 MiB -> 312.47 MiB (0.00%)

aliemjay commented:
Added a comment about rigid aliase types and implemented an [alternative approach](https://github.com/rust-lang/rust/pull/119084#discussion_r1430389025)

@rusbot review

compiler-errors commented:
r=me

aliemjay commented:
@bor r=compiler-errors  rollup=never

aliemjay commented:
@bors r=compiler-errors rollup=never

bors commented:
:pushpin: Commit 27e964dbfb9907970e96bd40429563449000bfbe has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 27e964dbfb9907970e96bd40429563449000bfbe -->
<!-- homu: {"type":"Approved","sha":"27e964dbfb9907970e96bd40429563449000bfbe","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 27e964dbfb9907970e96bd40429563449000bfbe with merge 57ad5058d643d06c0e76bb85442ff9244d966f34...
<!-- homu: {"type":"BuildStarted","head_sha":"27e964dbfb9907970e96bd40429563449000bfbe","merge_sha":"57ad5058d643d06c0e76bb85442ff9244d966f34"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7266551428/job/19801889251)
Approved by: compiler-errors
Pushing 57ad5058d643d06c0e76bb85442ff9244d966f34 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7266551428/job/19801889251"},"merge_sha":"57ad5058d643d06c0e76bb85442ff9244d966f34"} -->

nnethercote commented:
@aliemjay: how did you find this improvement? Profiling, looking through the code, something else? I'm curious :)

rust-timer commented:
Finished benchmarking commit (57ad5058d643d06c0e76bb85442ff9244d966f34): [comparison URL](https://perf.rust-lang.org/compare.html?start=3a539c08891a422d34beb32d2556a08ea0357b7d&end=57ad5058d643d06c0e76bb85442ff9244d966f34&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=3a539c08891a422d34beb32d2556a08ea0357b7d&end=57ad5058d643d06c0e76bb85442ff9244d966f34&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.9% | [-2.9%, -0.2%] | 8     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.9% | [-2.9%, -0.2%] | 8     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3a539c08891a422d34beb32d2556a08ea0357b7d&end=57ad5058d643d06c0e76bb85442ff9244d966f34&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.5%  | [3.5%, 3.5%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.6%, -0.5%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.6% | [-0.6%, -0.5%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=3a539c08891a422d34beb32d2556a08ea0357b7d&end=57ad5058d643d06c0e76bb85442ff9244d966f34&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.7% | [-1.8%, -1.5%] | 4     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.7% | [-1.8%, -1.5%] | 4     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=3a539c08891a422d34beb32d2556a08ea0357b7d&end=57ad5058d643d06c0e76bb85442ff9244d966f34&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.0% | [0.0%, 0.0%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

**Bootstrap**: 673.86s -> 671.988s (-0.28%)
**Artifact size**: 312.43 MiB -> 312.43 MiB (0.00%)

aliemjay commented:
> @aliemjay: how did you find this improvement? Profiling, looking through the code, something else? I'm curious :)

by callgrind profiles, looking at code that I have some expertise in so that I can easily catch optimazation opportunities.
